rte/kernels-openacc/mo_rte_solver_kernels.F90

```diff
@@ -198,7 +196,7 @@ subroutine lw_solver_noscat(ncol, nlay, ngpt, top_at_1, D, weight,
     !
     ! Surface reflection and emission
     !
-    !$acc parallel loop collapse(2)
+    !$acc parallel loop collapse(2) no_create(gpt_Jac)
```

Comment:
Mostly just trying to understand the code. does `no_create(gpt_Jac)` have a performance benefit?

Reply:
`no_create` wouldn't create the `gpt_Jac` array on the GPU if it has not been allocated. So like `present if (acc_present(gpt_Jac))`. So in case we don't need the Jacobians, we won't allocate the said array on the device in this kernel.

Reply:
Basically, `gpt_Jac` is an optional argument somewhere higher up in the calls stack and so it may not exist at all on the CPU.

OK, I am less sure how to do this in OpenMP offload or even what the OpenMP runtime will do. Normally, implicit mapping rules would say to map the array even if not explictly mapped. But here you have a call path that is only executed, if `gpt_Jac` exists. https://github.com/earth-system-radiation/rte-rrtmgp/blob/026cd9d81ce228f2f66db28c0f47c0cec89a7ab4/rte/kernels-openacc/mo_rte_solver_kernels.F90#L208-L210

I am not sure how to do this with OpenMP offload or maybe it is fine that way it is. Let me ask around.

Reply:
I don't think the OpenMP offload version of this code will work. I remember talking to @RobertPincus about this because we found some odd bugs in the Cray compiler. We drafted something, but never committed the changes.

I think we need to do something like this
```
 if(do_Jacobians)
   !$acc parallel loop collapse(2)
    !$omp target teams distribute parallel do simd collapse(2)
    do igpt = 1, ngpt
      do icol = 1, ncol
        !
        ! Surface albedo, surface source function
        !
        radn_up   (icol,sfc_level,igpt) = radn_dn(icol,sfc_level,igpt)*(1._wp - sfc_emis(icol,igpt)) + &
                                          sfc_src(icol,          igpt)*         sfc_emis(icol,igpt)
        gpt_Jac(icol,sfc_level,igpt) = sfc_srcJac(icol,        igpt)*         sfc_emis(icol,igpt)
      end do
    end do
else
   !$acc parallel loop collapse(2)
    !$omp target teams distribute parallel do simd collapse(2)
    do igpt = 1, ngpt
      do icol = 1, ncol
        !
        ! Surface albedo, surface source function
        !
        radn_up   (icol,sfc_level,igpt) = radn_dn(icol,sfc_level,igpt)*(1._wp - sfc_emis(icol,igpt)) + &
                                          sfc_src(icol,          igpt)*         sfc_emis(icol,igpt)
      end do
    end do
end
```
but its ugly to have to have duplicated code.

Reply:
@naromero77 Seems like we should just have separate kernels: 
```suggestion
   !$acc parallel loop collapse(2)
    !$omp target teams distribute parallel do simd collapse(2)
    do igpt = 1, ngpt
      do icol = 1, ncol
        !
        ! Surface albedo, surface source function
        !
        radn_up   (icol,sfc_level,igpt) = radn_dn(icol,sfc_level,igpt)*(1._wp - sfc_emis(icol,igpt)) + &
                                          sfc_src(icol,          igpt)*         sfc_emis(icol,igpt)
      end do
    end do

 if(do_Jacobians)
   !$acc parallel loop collapse(2)
    !$omp target teams distribute parallel do simd collapse(2)
    do igpt = 1, ngpt
      do icol = 1, ncol
        !
        ! Surface albedo, surface source function
        !
        gpt_Jac(icol,sfc_level,igpt) = sfc_srcJac(icol,        igpt)*         sfc_emis(icol,igpt)
      end do
    end do
end if 
```

Reply:
yes, that would be safer IMO.

Reply:
I think its OK for this PR to be merged, but I still have concerns that this section of code may cause problems in the future for OpenMP offload.

Reply:
@naromero77 Feel free to open an issue, a PR, or both. 

Note that we've discussed separating the Jacobian (= optional arguments) kernels for other reasons.